File: mobile_projects/IoTAI-Demo-Android/app/src/main/java/com/ignitmeetup/iotai_demo_android/ContentManager.kt
package com.ignitmeetup.iotai_demo_android

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.tensorflow.lite.Interpreter
import java.nio.ByteBuffer
import java.nio.ByteOrder

class ContentManager(
    private val interpreter: Interpreter
) : ViewModel() {

    private val _currentProbability = MutableStateFlow<Float?>(null)
    val currentProbability: StateFlow<Float?> = _currentProbability

    private val _maintenanceNeeded = MutableStateFlow(false)
    val maintenanceNeeded: StateFlow<Boolean> = _maintenanceNeeded

    private val _probabilityHistory = MutableStateFlow<List<Float>>(emptyList())
    val probabilityHistory: StateFlow<List<Float>> = _probabilityHistory

    private val _stepCurrentValue = MutableStateFlow("")
    val stepCurrentValue: StateFlow<String> = _stepCurrentValue

    private val _tempCurrentValue = MutableStateFlow("")
    val tempCurrentValue: StateFlow<String> = _tempCurrentValue

    private val _humiCurrentValue = MutableStateFlow("")
    val humiCurrentValue: StateFlow<String> = _humiCurrentValue

    private val _pressCurrentValue = MutableStateFlow("")
    val pressCurrentValue: StateFlow<String> = _pressCurrentValue

    // Normalization constants
    private val tempMean: Float = 21.1f    // Mean temperature (°C)
    private val tempStd: Float = 1.1f      // Standard deviation of temperature (°C)
    private val humidMean: Float = 50.0f   // Mean humidity (%)
    private val humidStd: Float = 5.0f     // Standard deviation of humidity (%)
    private val pressMean: Float = 300.0f  // Mean pressure (psi)
    private val pressStd: Float = 10.0f    // Standard deviation of pressure (psi)

    // HVAC input data class
    data class HVACInput(
        val temperature: Float,  // °C
        val humidity: Float,     // %
        val pressure: Float      // psi
    )

    // Start the prediction simulation
    fun startPredicting() {
        viewModelScope.launch(Dispatchers.IO) {
            try {
                val numberOfSteps = 70

                val sequence = generateTrendSequence(numberOfSteps)

                for ((index, input) in sequence.withIndex()) {
                    // Normalize input and prepare TFLite input buffer
                    val inputBuffer = normalizeInput(input) // Input preparation logic
                    val output = Array(1) { FloatArray(1) } // Matches [1, 1] shape
                    interpreter.run(inputBuffer, output)
                    val probability = output[0][0].coerceIn(0f, 1f) // Ensure value is between 0 and 1

                    // Log for debugging
                    println(
                        "Step ${index + 1}: Temp = %.2f°C, Hum = %.2f%%, Press = %.2f psi -> Probability: %.3f"
                            .format(input.temperature, input.humidity, input.pressure, probability)
                    )

                    // Update UI state on the main thread
                    withContext(Dispatchers.Main) {
                        _currentProbability.value = probability
                        _maintenanceNeeded.value = probability > 0.5f
                        _probabilityHistory.value += probability

                        _stepCurrentValue.value = "Step = ${index + 1} / $numberOfSteps"
                        _tempCurrentValue.value = "Temp = %.2f °C".format(input.temperature)
                        _humiCurrentValue.value = "Hum = %.2f %%".format(input.humidity)
                        _pressCurrentValue.value = "Press = %.2f psi".format(input.pressure)
                    }

                    delay(500) // 0.5 second delay to simulate real-time data
                }
            } catch (e: Exception) {
                println("Prediction failed: $e")
            }
        }
    }

    // Generate a sequence of HVAC inputs
    private fun generateTrendSequence(steps: Int): List<HVACInput> {
        val sequence = mutableListOf<HVACInput>()
        val totalSteps = maxOf(steps, 20)
        val postMaintenanceSteps = minOf(20, totalSteps / 3)
        val stableSteps = (totalSteps - postMaintenanceSteps) / 2
        val deteriorationSteps = totalSteps - stableSteps - postMaintenanceSteps

        val startTemp = 21.1f
        val startHumidity = 50.0f
        val startPressure = 300.0f

        val deterioratedTemp = startTemp + 5f
        val deterioratedHumidity = startHumidity + 15f
        val deterioratedPressure = startPressure - 100f

        for (i in 0 until totalSteps) {
            when {
                i < stableSteps -> {
                    val temp = startTemp + (-0.5f..0.5f).random()
                    val hum = startHumidity + (-1f..1f).random()
                    val press = startPressure + (-2f..2f).random()
                    sequence.add(HVACInput(temp, hum, press))
                }

                i < stableSteps + deteriorationSteps -> {
                    val fraction = (i - stableSteps).toFloat() / (deteriorationSteps - 1)
                    val temp =
                        startTemp + fraction * (deterioratedTemp - startTemp) + (-0.5f..0.5f).random()
                    val hum =
                        startHumidity + fraction * (deterioratedHumidity - startHumidity) + (-1f..1f).random()
                    val press =
                        startPressure + fraction * (deterioratedPressure - startPressure) + (-2f..2f).random()
                    sequence.add(HVACInput(temp, hum, press))
                }

                else -> {
                    val temp = startTemp + (-0.5f..0.5f).random()
                    val hum = startHumidity + (-1f..1f).random()
                    val press = startPressure + (-2f..2f).random()
                    sequence.add(HVACInput(temp, hum, press))
                }
            }
        }
        return sequence
    }

    // Normalize input for TFLite model
    private fun normalizeInput(input: HVACInput): ByteBuffer {
        val buffer = ByteBuffer.allocateDirect(3 * 4) // 3 floats, 4 bytes each
        buffer.order(ByteOrder.nativeOrder())
        buffer.putFloat((input.temperature - tempMean) / tempStd)
        buffer.putFloat((input.humidity - humidMean) / humidStd)
        buffer.putFloat((input.pressure - pressMean) / pressStd)
        buffer.rewind()
        return buffer
    }

    // Extension function to generate random Float in range
    private fun ClosedRange<Float>.random(): Float =
        (endInclusive - start) * kotlin.random.Random.nextFloat() + start
}

File: mobile_projects/IoTAI-Demo-iOS/IoTAI-Demo-iOS/ContentManager.swift
//
//  ContentManager.swift
//  IoTAI-Demo-iOS
//
//  Created by Matija Kruljac on 11.03.2025.
//

import CoreML
import Foundation

class ContentManager: ObservableObject {

    @Published var currentProbability: Float?
    @Published var maintenanceNeeded: Bool = false
    @Published var probabilityHistory: [Float] = []

    @Published var stepCurrentValue: String = ""
    @Published var tempCurrentValue: String = ""
    @Published var humiCurrentValue: String = ""
    @Published var pressCurrentValue: String = ""

    // MARK: - Normalization Parameters

    /// Normalization constants based on actual training data statistics
    let tempMean: Float = 21.1    // Mean temperature (°C)
    let tempStd: Float = 1.1      // Standard deviation of temperature (°C)
    let humidMean: Float = 50.0   // Mean humidity (%)
    let humidStd: Float = 5.0     // Standard deviation of humidity (%)
    let pressMean: Float = 300.0  // Mean pressure (psi)
    let pressStd: Float = 10.0    // Standard deviation of pressure (psi)

    // MARK: - Prediction Logic

    /// Runs the HVAC maintenance prediction simulation
    func startPredicting() {
        DispatchQueue.global(qos: .background).async {
            do {
                // Initialize the Core ML model
                let model = try HVACModel(configuration: MLModelConfiguration())

                let numberOfSteps = 70

                // Generate a sequence of 'numberOfSteps' inputs simulating a trend
                let sequence = self.generateTrendSequence(steps: numberOfSteps)

                // Process each input with a 0.5 second delay
                for (index, input) in sequence.enumerated() {
                    let multiArray = self.hvacInputToMultiArray(input)
                    let modelInput = HVACModelInput(input_1: multiArray)
                    let prediction = try model.prediction(input: modelInput)

                    // Extract the probability from the model's output
                    let probability = prediction.Identity[0].floatValue

                    // Print sensor values and prediction result
                    print("Step \(index + 1): Temp = \(String(format: "%.2f", input.temperature))°C, " +
                          "Hum = \(String(format: "%.2f", input.humidity))%, " +
                          "Press = \(String(format: "%.2f", input.pressure)) psi -> " +
                          "Probability: \(String(format: "%.3f", probability))")

                    // Update properties on the main thread
                    DispatchQueue.main.async {
                        self.currentProbability = probability
                        self.maintenanceNeeded = probability > 0.5
                        self.probabilityHistory.append(probability)

                        self.stepCurrentValue = "Step = \(index + 1)/\(numberOfSteps)"
                        self.tempCurrentValue = "Temp = \(String(format: "%.2f", input.temperature)) °C"
                        self.humiCurrentValue = "Hum = \(String(format: "%.2f", input.humidity)) %"
                        self.pressCurrentValue = "Press = \(String(format: "%.2f", input.pressure)) psi"
                    }

                    // Simulate real-time data emission with a 0.5 second delay
                    Thread.sleep(forTimeInterval: 0.5)
                }
            } catch {
                print("Prediction failed: \(error)")
            }
        }
    }

    // MARK: - Data Structures

    /// Struct to represent HVAC input features
    struct HVACInput {
        let temperature: Float  // °C
        let humidity: Float     // %
        let pressure: Float     // psi
    }

    // MARK: - Helper Functions

    /// Generates a sequence of HVAC inputs simulating a trend toward maintenance and recovery
    /// - Parameter steps: Number of steps in the sequence (minimum 20 to accommodate post-maintenance phase)
    /// - Returns: Array of HVACInput structs
    private func generateTrendSequence(steps: Int) -> [HVACInput] {
        var sequence: [HVACInput] = []
        let totalSteps = max(steps, 20) // Ensure at least 20 steps for meaningful phases
        let postMaintenanceSteps = min(20, totalSteps / 3) // Last 15–20 steps are post-maintenance
        let stableSteps = (totalSteps - postMaintenanceSteps) / 2 // First phase: stable operation
        let deteriorationSteps = totalSteps - stableSteps - postMaintenanceSteps // Middle phase: deterioration

        // Starting values close to the training means
        let startTemp: Float = 21.1
        let startHumidity: Float = 50.0
        let startPressure: Float = 300.0

        // Deterioration values (temp up 5°C, humidity up 15%, pressure down 100 psi)
        let deterioratedTemp = startTemp + 5
        let deterioratedHumidity = startHumidity + 15
        let deterioratedPressure = startPressure - 100

        for i in 0..<totalSteps {
            if i < stableSteps {
                // Phase 1: Stable operation with small random noise
                let temp = startTemp + Float.random(in: -0.5...0.5)
                let hum = startHumidity + Float.random(in: -1...1)
                let press = startPressure + Float.random(in: -2...2)
                sequence.append(HVACInput(temperature: temp, humidity: hum, pressure: press))
            } else if i < stableSteps + deteriorationSteps {
                // Phase 2: Gradual deterioration toward deteriorated values
                let fraction = Float(i - stableSteps) / Float(deteriorationSteps - 1)
                let temp = startTemp + fraction * (deterioratedTemp - startTemp) + Float.random(in: -0.5...0.5)
                let hum = startHumidity + fraction * (deterioratedHumidity - startHumidity) + Float.random(in: -1...1)
                let press = startPressure + fraction * (deterioratedPressure - startPressure) + Float.random(in: -2...2)
                sequence.append(HVACInput(temperature: temp, humidity: hum, pressure: press))
            } else {
                // Phase 3: Post-maintenance - return to normal conditions
                let temp = startTemp + Float.random(in: -0.5...0.5)
                let hum = startHumidity + Float.random(in: -1...1)
                let press = startPressure + Float.random(in: -2...2)
                sequence.append(HVACInput(temperature: temp, humidity: hum, pressure: press))
            }
        }

        return sequence
    }

    /// Converts HVACInput to a normalized MLMultiArray for model input
    /// - Parameter input: HVACInput struct with raw sensor values
    /// - Returns: Normalized MLMultiArray
    private func hvacInputToMultiArray(_ input: HVACInput) -> MLMultiArray {
        // shape: [1, 3] => This tells MLMultiArray to create a 2D array with 1 row and 3 columns
        // Example: [ [temperature, humidity, pressure] ]
        let multiArray = try! MLMultiArray(shape: [1, 3], dataType: .float32)

        // Normalize each feature using (value - mean) / std
        let normalizedTemp = (input.temperature - tempMean) / tempStd
        let normalizedHumidity = (input.humidity - humidMean) / humidStd
        let normalizedPressure = (input.pressure - pressMean) / pressStd
        
        // Assign normalized values to the multi-array
        multiArray[0] = NSNumber(value: normalizedTemp)
        multiArray[1] = NSNumber(value: normalizedHumidity)
        multiArray[2] = NSNumber(value: normalizedPressure)

        return multiArray
    }
}
